File: modules/kakao-user/android/src/main/java/expo/modules/kakaouser/KakaoUserModule.kt
package expo.modules.kakaouser

import com.kakao.sdk.auth.AuthApiClient
import com.kakao.sdk.auth.model.OAuthToken
import com.kakao.sdk.user.UserApiClient
import expo.modules.kakaocore.KakaoErrorUtil
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class KakaoUserModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("KakaoUser")

      AsyncFunction("login") { promise: Promise ->
      val context =
              appContext.activityProvider?.currentActivity
                      ?: run {
                        return@AsyncFunction promise.resolve(
                                KakaoLoginResult.failure(
                                        KakaoErrorUtil.formatError<KakaoUserModule>(
                                                KakaoErrorUtil.CommonErrors.ACTIVITY_NOT_FOUND
                                        )
                                )
                        )
                      }

      val callback: (OAuthToken?, Throwable?) -> Unit = { oauthToken, error ->
        when {
          error != null ->
                  promise.resolve(
                          KakaoLoginResult.failure(
                                  KakaoErrorUtil.formatError<KakaoUserModule>(
                                          error.message ?: KakaoErrorUtil.CommonErrors.UNKNOWN_ERROR
                                  )
                          )
                  )
          oauthToken == null ->
                  promise.resolve(
                          KakaoLoginResult.failure(
                                  KakaoErrorUtil.formatError<KakaoUserModule>(
                                          KakaoErrorUtil.CommonErrors.TOKEN_NOT_FOUND
                                  )
                          )
                  )
          else -> {
            val token = AuthTokenModel.from(oauthToken)
            promise.resolve(KakaoLoginResult.success(token))
          }
        }
      }

      if (UserApiClient.instance.isKakaoTalkLoginAvailable(context)) {
        UserApiClient.instance.loginWithKakaoTalk(context = context, callback = callback)
      } else {
        UserApiClient.instance.loginWithKakaoAccount(context = context, callback = callback)
      }
    }
      AsyncFunction("logout") { promise: Promise ->
          UserApiClient.instance.logout { it ->
              when {
                  it != null ->  promise.resolve(
                      KakaoLogoutResult.failure(
                          KakaoErrorUtil.formatError<KakaoUserModule>(it.message?: KakaoErrorUtil.CommonErrors.UNKNOWN_ERROR)
                      )
                  )
                  else ->
                      promise.resolve(
                          KakaoLogoutResult.success()
                      )
              }
          }
      }
      AsyncFunction("unlink") {promise:Promise ->
          UserApiClient.instance.unlink { it ->
              when {
                  it != null ->  promise.resolve(
                      KakaoLogoutResult.failure(
                          KakaoErrorUtil.formatError<KakaoUserModule>(it.message?: KakaoErrorUtil.CommonErrors.UNKNOWN_ERROR)
                      )
                  )
                  else ->
                      promise.resolve(
                          KakaoLogoutResult.success()
                      )
              }
          }
      }
      AsyncFunction("isLogined") {promise: Promise->
          when {
              AuthApiClient.instance.hasToken() -> UserApiClient.instance.accessTokenInfo {_, error ->
                  when {
                      error != null -> promise.resolve(
                          KakaoIsLogined.error(KakaoErrorUtil.formatError<KakaoUserModule>(
                              error.message?: "Login check Error"
                          ))
                      )
                      else -> promise.resolve(
                          KakaoIsLogined.loggedIn()
                      )
                  }
              }
              else -> promise.resolve(
                  KakaoIsLogined.notLoggedIn()
              )
          }
      }
      AsyncFunction("getAccessToken") { promise:Promise ->
          UserApiClient.instance.accessTokenInfo { accessToken, error ->
              when {
                  error != null -> promise.resolve(
                      KakaoAccessTokenResult.failure(
                          KakaoErrorUtil.formatError<KakaoUserModule>(
                              error.message?:"Access Token Check Error"
                          )
                      )
                  )
                  accessToken == null -> promise.resolve(
                      KakaoAccessTokenResult.failure(
                          KakaoErrorUtil.formatError<KakaoUserModule>(
                              KakaoErrorUtil.CommonErrors.TOKEN_NOT_FOUND
                          )
                      )
                  )
                  else -> promise.resolve(
                      KakaoAccessTokenResult.success(
                          AccessTokenInfoModel.from(accessToken)
                      )
                  )
              }
          }
      }
  }
}

File: modules/kakao-user/ios/KakaoUserModule.swift
import ExpoModulesCore
import KakaoCore
import KakaoSDKAuth
import KakaoSDKCommon
import KakaoSDKUser

public class KakaoUserModule: Module {
  public func definition() -> ModuleDefinition {
    Name("KakaoUser")

    AsyncFunction("login") { (promise: Promise) in
      guard (try? KakaoSDK.shared.appKey()) != nil else {
        let result = KakaoLoginResultModel(
          error: KakaoErrorUtil.formatError(
            from: Self.self, errorMessage: KakaoErrorUtil.CommonErrors.sdkNotInitialized
          ))
        return promise.resolve(result)
      }

      let callback = { (oauthToken: OAuthToken?, error: Error?) in
        if let error = error {
          let result = KakaoLoginResultModel(
            error: KakaoErrorUtil.formatError(
              from: Self.self, errorMessage: error.localizedDescription
            ))
          promise.resolve(result)
        } else if let token = oauthToken {
          let tokenModel = AuthTokenModel.from(token)
          let result = KakaoLoginResultModel(
            success: true,
            token: tokenModel
          )
          promise.resolve(result)
        } else {
          let result = KakaoLoginResultModel(
            error: KakaoErrorUtil.formatError(
              from: Self.self, errorMessage: KakaoErrorUtil.CommonErrors.tokenNotFound
            )
          )
          promise.resolve(result)
        }
      }

      if UserApi.isKakaoTalkLoginAvailable() {
        UserApi.shared.loginWithKakaoTalk(completion: callback)
      } else {
        UserApi.shared.loginWithKakaoAccount(completion: callback)
      }
    }

    AsyncFunction("logout") { (promise: Promise) in
      UserApi.shared.logout { error in
        if let error {
          promise.resolve(
            KakaoLogoutResult(
              error: KakaoErrorUtil.formatError(from: Self.self, errorMessage: error.localizedDescription)
            )
          )
        } else {
          promise.resolve(
            KakaoLogoutResult(success: true)
          )
        }
      }
    }

    AsyncFunction("unlink") { (promise: Promise) in
      UserApi.shared.unlink { error in
        if let error {
          promise.resolve(
            KakaoLogoutResult(
              error: KakaoErrorUtil.formatError(from: Self.self, errorMessage: error.localizedDescription)
            )
          )
        } else {
          promise.resolve(
            KakaoLogoutResult(success: true)
          )
        }
      }
    }

    AsyncFunction("isLogined") { (promise: Promise) in
      if AuthApi.hasToken() {
        UserApi.shared.accessTokenInfo { _, error in
          if let error {
            promise.resolve(
              KakaoIsLogined(
                error: KakaoErrorUtil.formatError(from: Self.self, errorMessage: error.localizedDescription)
              )
            )
          } else {
            promise.resolve(
              KakaoIsLogined(isLogined: true)
            )
          }
        }
      } else {
        promise.resolve(KakaoIsLogined())
      }
    }

    AsyncFunction("getAccessToken") { (promise: Promise) in
      UserApi.shared.accessTokenInfo { accessToken, error in
        if let error {
          promise.resolve(
            KakaoAccessTokenResult(error: KakaoErrorUtil.formatError(from: Self.self, errorMessage: error.localizedDescription)
            )
          )
        } else if let accessToken {
          promise.resolve(
            KakaoAccessTokenResult(
              accessTokenInfo: AccessTokenInfoModel.from(accessToken)
            )
          )
        } else {
          promise.resolve(
            KakaoAccessTokenResult(error: KakaoErrorUtil.formatError(from: Self.self, errorMessage: KakaoErrorUtil.CommonErrors.tokenNotFound)
            )
          )
        }
      }
    }
  }
}
